[FEATURE] Récupération des données API via le code de vérification (PIX-773)

P-Jeremy opened the pull request:
## :unicorn: Problème
Dans le contexte du partage du contrôle du certificat, nous souhaitons qu'une personne ayant réussis sa certification puisse la partager avec une / des personnes tierces.
Ceci se fait via un "code de vérification", **prochainement** visible sur chaque certificats des candidats ayant réussi leur test de certification.
La personne à qui est adressée le code, peut alors aller, **prochainement** sur une page affichant un formulaire, dans lequel il faut remplir un "code de vérification" ainsi qu'un "score pix".
Si elle renseigne le bon " pix score" et le bon "code de vérif" alors le certificat s'affichera (en omettant le commentaire du jury à destination du candidat).

## :robot: Solution
Cette PR permet donc de renvoyer toutes les informations nécessaires à l'affichage d'un certificat lorsqu'on fournit un "pix score" ainsi qu'un "code de vérification".

## :rainbow: Remarques
Durant cette PR 2 nouvelles notions ont été nommée : 
- le `ShareableCertificate` : qui correspond aux informations affichées sur le certificat qu'a partagé le candidat et vu par une autre personne (un potentiel recruteur ou autre).
- le `PrivateCertificate` : qui correspond aux informations affichées sur le certificat d'un candidat s'était connecté sur mon-pix et regardant son certificat depuis l'onglet "Mes certifications".
Il contient en plus du `ShareableCertificate` le commentaire du Jury Pix ainsi que le code de vérification.

## :100: Pour tester
Rien n'est rajouté d'un point de vue front.
Faire un POST sur `/api/shared-certifications` avec un `verificationCode` et `pixScore`.
Pour le `verificationCode`, dans nos seeds on en trouve dans n'importe quel `certification-courses` (le mieux reste de rentrer dans une session de certification*).
Pour le `pixScore` il est dans les `assessment-results` (les `assessment-results` ont une clé étrangère vers les `assessements` qui ont une clé étrangère vers les `certification-courses`). 

Aller dans une session de certification : 
- Se connecter sur app avec n'importe quel user certifiable, par exemple, `certif1@example.net` (`pix123`) 
- Aller dans l'onglet "Certification" et rentrer dans la session numéro `2`, via le candidat `Anne`, `Success`, née le `01-01-2000`. Le code de session est `ANNE02`.
🎉  votre `certification-course` avec son `certificationCode` viens d'être créé !

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr1752.review.pix.fr
- Orga: https://orga-pr1752.review.pix.fr
- Certif: https://certif-pr1752.review.pix.fr
- Admin: https://admin-pr1752.review.pix.fr
- API: https://api-pr1752.review.pix.fr/api/

 Please check it out!

P-Jeremy committed:
Add getSharedCertification in certifications repo

Anne-Gaelle-S committed:
Add getSharedCertification usecase

Anne-Gaelle-S committed:
Add serializerForSharing

Anne-Gaelle-S committed:
Add getSharedCertif in controller

Giroteti committed:
Centralize common getCertificate() behavior

Giroteti committed:
Distinguish Private- from Shareable-Certificate

Rename Certification into Certificate
PrivateCertificate has examiner comment
ShareableCertificate has no examiner comment

P-Jeremy committed:
Adding predictible verificationCode to seed

P-Jeremy committed:
Switch unauthorized by notFound error

Anne-Gaelle-S committed:
Remove pixScore from request's payload

Anne-Gaelle-S commented on api/lib/domain/usecases/certificate/get-shareable-certificate.js:
```diff
@@ -1,15 +1,18 @@
 }) {
-  const getBaseCertificate = async () => certificationRepository.getShareableCertificateByVerificationCode({ verificationCode });
-  const isAccessToCertificateAuthorized = (certification) => certification.pixScore === pixScore;
+  const splitInputCode = verificationCode.toUpperCase().split('-');
+  const certificationVerificationCode = splitInputCode.slice(0,2).join('-');
+  const pixScore = splitInputCode[2];
```
```suggestion
  const splitInputCode = verificationCode && verificationCode.toUpperCase().split('-');
  const certificationVerificationCode = splitInputCode && splitInputCode.slice(0,2).join('-');
  const pixScore = splitInputCode && splitInputCode[2];
```

laura-bergoens commented on api/lib/domain/usecases/get-shared-certification.js:
```diff
@@ -0,0 +1,59 @@
+const ResultCompetenceTree = require('../models/ResultCompetenceTree');
+const { UserNotAuthorizedToAccessEntity } = require('../errors');
+
+async function _getsCompetenceMarksAndAssessmentResultId({ certificationId, assessmentResultRepository }) {
```
pourquoi y'a un s dans gets

laura-bergoens reviewed (approved):
Globalement OK, sauf :
- pourquoi un POST et pas un GET ?
- pas testé fonctionnellement

P-Jeremy commented:
> Globalement OK, sauf :
> 
> * pourquoi un POST et pas un GET ?
> * pas testé fonctionnellement

@laura-bergoens  Pour ne pas avoir à passer le code et le score de l'utilisateur dans l'url de la requète

Anne-Gaelle-S committed:
Add /shared-certifications route

Anne-Gaelle-S commented:
Vu avec @yahyayouness : pour récupérer des données le verbe approprié c'est GET.
Ensuite pour ne pas passer le code de vérif en paramètre (et donc empêcher des tiers de le récupérer) on peut le passer dans les headers de la requête.
Donc je change le POST pour un GET :)

Anne-Gaelle-S commented:
Retournement de situation, finalement ça sera un POST ! 
> L’utilisation d’un header custom est contradictoire avec une mise en cache par ailleurs. Ou plutôt, ça crée même un risque qu’un cache (qui n’a pas été correctement prévenu que ce header était important avec une directive Vary) garde la réponse et la serve à quelqu’un d’autre qui ne mettrait pas le header.
Voir https://1024pix.slack.com/archives/C013ZKZPT5Y/p1597925818002200

Anne-Gaelle-S committed:
Change GET to POST verb
